wasm-chain-integration/src/v1/mod.rs

```diff
@@ -0,0 +1,845 @@
+                    v0::host::get_receive_owner(memory, stack, self.receive_ctx.owner())
+                }
+            }?,
+            ImportFunc::InitOnly(InitOnlyFunc::GetInitOrigin) => {
+                bail!("Not implemented for receive.");
+            }
+        }
+        Ok(None)
+    }
+}
+
+/// Parameter whose ownership is tracked statically.
+pub type ParameterRef<'a> = &'a [u8];
+/// Parameter whose ownership is tracked dynamically.
+/// This is needed, for example, when execution passes through Haskell and Rust.
+/// Ideally this would be Arc<[u8]> but then this cannot be passed via the FFI
+/// boundary directly since [u8] is not Sized. To avoid a third type we settle
+/// on Vec<u8>.
+pub type ParameterVec = Vec<u8>;
+
+/// Invokes an init-function from a given artifact
+pub fn invoke_init<R: RunnableCode>(
+    artifact: impl Borrow<Artifact<ProcessedImports, R>>,
+    amount: u64,
+    init_ctx: impl v0::HasInitContext,
+    init_name: &str,
+    param: ParameterRef,
+    energy: u64,
+) -> ExecResult<InitResult> {
+    let mut host = InitHost {
+        energy: InterpreterEnergy {
+            energy,
+        },
+        activation_frames: constants::MAX_ACTIVATION_FRAMES,
+        logs: v0::Logs::new(),
+        state: v0::State::new(None),
+        return_value: Vec::new(),
+        parameters: vec![param],
+        init_ctx,
+    };
+
+    let result = artifact.borrow().run(&mut host, init_name, &[Value::I64(amount as i64)]);
+    process_init_result(host, result)
+}
+
+fn process_init_result<Param, Ctx>(
+    host: InitHost<Param, Ctx>,
+    result: machine::RunResult<ExecutionOutcome<NoInterrupt>>,
+) -> ExecResult<InitResult> {
+    match result {
+        Ok(ExecutionOutcome::Success {
+            result,
+            ..
+        }) => {
+            let remaining_energy = host.energy.energy;
+            // process the return value.
+            // - 0 indicates success
+            // - positive values are a protocol violation, so they lead to a runtime error
+            // - negative values lead to a rejection with a specific reject reason.
+            if let Some(Value::I32(n)) = result {
+                if n == 0 {
+                    Ok(InitResult::Success {
+                        state: host.state,
+                        logs: host.logs,
+                        return_value: host.return_value,
+                        remaining_energy,
+                    })
+                } else {
+                    Ok(InitResult::Reject {
+                        reason: reason_from_wasm_error_code(n)?,
+                        return_value: host.return_value,
+                        remaining_energy,
+                    })
+                }
+            } else {
+                bail!("Wasm module should return a value.")
+            }
+        }
+        Ok(ExecutionOutcome::Interrupted {
+            reason,
+            config: _,
+        }) => match reason {},
```

Comment:
Magic! Init functions cannot interrupt.

The way interrupts are done is that a host function can either return normally, or it can trigger an interrupt (return a given interrupt type). That is, the return value of a host call is
```machine::RunResult<Option<Self::Interrupt>>```

We set the `Interrupt` type to be the empty type for the init method, which means that this case is impossible so we do not have to handle it.

Reply:
Ah okay. Nice 👍